scylla/src/cluster/metadata.rs

```
@@ -452,6 +455,7 @@ impl MetadataReader {
             host_filter: host_filter.clone(),
             initial_known_nodes,
             control_connection_repair_requester,
+            #[cfg(feature = "metrics")]
             metrics,
         })
     }
```

@muzarski @wprzytula I'm looking at the commit that puts metrics behind crate future, and I'm wondering:
maybe it is better to restrict those `cfg`'s to the metrics module, by providing alternative implementation which will be active when `cfg(not(feature = "metrics"))`. This implementation would have no fields, but all the same methods, which would just be empty.

Why? The reasons to put metrics behind future are, afaik:
- Performance
- Making histogram dependency optional

Doing what I proposed still achieves those goals, but also remove a lot of conditional compilation flags from other files in the crate, which I think is a great plus for code readability.

---

The method on `Session` that retrieves the metrics would still be under `#[cfg(feature = "metrics")]` of course.

---

> @muzarski @wprzytula I'm looking at the commit that puts metrics behind crate future, and I'm wondering: maybe it is better to restrict those `cfg`'s to the metrics module, by providing alternative implementation which will be active when `cfg(not(feature = "metrics"))`. This implementation would have no fields, but all the same methods, which would just be empty.

I understand your idea. The only thing I don't get: "but all the same methods, which would just be **empty**". How can a method be empty if it returns something other than `()`? Apart from that, your suggestion sounds reasonable.

---

This artifical `Metrics` type only needs `inc_*`, `dec_*`, and `log_*` methods because those are used to record data.
All such methods return `()`, apart from `log_query_latency`, which returns `Result`, but keeping the `MetricsError` (maybe as pub(crate) if feature is disabled) is not a problem.
Alternatively it could ignore the error - both call site of this method already do that by using `let _ = ...`.

---

This sounds like extremely premature optimization. Unless we have proof, or at least reasonable suspicion, that it impacts performance in any significant way, I don't agree to sacrifice readability for optimizations.

---

Another thing I'm against is having metrics structs, passing them and calling methods on them, even if they are fully non-op. This is nonintuitive and hides the fact that the feature is needed for those operations to really perform meaningful stuff. WDYT @Lorak-mmk ?

---

That's a much more reasonable objection!
How I see it:
- I believe the runtime impact will not be meaningful, so we should focus on code quality.
- The code that interacts with metric will be in the source anyway. In fact there will be less of it if we remove the cfg lines.
- Connection / Session does not need to care what Metrics will do with the data provided to them. I see at as sort of a compile-time traits. You would not raise the same objections for the code having `Arc<dyn MetricsCollector>` and calling its methods, right?

---

> * You would not raise the same objections for the code having `Arc<dyn MetricsCollector>` and calling its methods, right?

In fact, I would. The difference is that we have `Option<Arc<dyn Policy>>` instead of `Arc<dyn Policy>` for run-time traits. If you agree to have metrics behind `Option`, then I would not raise this particular objection.

---

I really don't see the difference in this case. As far as I know we never had any reports about Arc's being a performance problem in the driver. Let's not do micro optimizations unless we have data showing they are useful.
Right now we also always clonde the metrics Arc - as we do with ClusterState Arc.
Please also note, that my proposed change can be reverted if it turns out that performance is a problem here - it will not be a breaking change.

---

Correct observation, both the change and reverting it are not API-breaking. Therefore, I suggest leaving this as-is and possibly addressing your suggestion after 1.0. WDYT?